### ISSUE TITLE
u8_clz function doesn't work when using GCC builtins

### ISSUE THREAD

**Issue by craigbarnes** [davidgarland/circa#9]
As demonstrated by the following:

```c
#include <assert.h>
#include "bits.h"

int main(void) {
    uint8_t nbits = u8_clz(0x01);
    printf("nbits: %hhu\n", nbits);
    assert(nbits == 7);
    return 0;
}
```
...which prints:

```
nbits: 31
a.out: test_bits.c:7: main: Assertion `nbits == 7' failed.
Aborted (core dumped)
```

Just delegating to `__builtin_clz` is always going to return the result for an `unsigned int`, so it's returning 7 plus the extra 24 padding 0-bits that get added by the implicit conversion.

**Comment by davidgarland** [comment]
OK, released the hotfix. Does this look good? https://github.com/davidgarland/circa/commit/390499649bd2d7d46aa9b1f164b35ae48a0ebc13

For the best portability, I decided to use `__builtin_clzll` since its guaranteed to be at least 64 bits (and therefore *is* exactly 64 bits on most platforms) and then subtracted any extra zeroes relative to the type being tested.

**Comment by craigbarnes** [comment]
> For the best portability, I decided to use `__builtin_clzll` since its guaranteed to be at least 64 bits (and therefore is exactly 64 bits on most platforms)

Yeah that should almost always work in practice, even though it's not guaranteed by the spec. The only possible issue I can see is that counting and then discarding 32 more bytes than is necessary might be slower under certain circumstances (e.g. if the builtin expands to a library routine instead of a single instruction).

**Comment by davidgarland** [comment]
In practice that won't ever happen, because everything is statically linked. Even in the upcoming rewrite I'm keeping the `bits.h` intrinsics as statically linked header functions to avoid any routines.

There may be some added cost associated with the subtraction, but this can likely be mitigated by doing some constant-folded if statements or ternary expressions instead of doing the subtraction. That said, it shouldn't be a very big performance difference, so I wouldn't call it a huge priority, but I'll get around to it before I release the rework.
